### PR TITLE
backfill: add memory accounting to the column/index backfiller

### LINKED ISSUE #46346: backfill: half of backfills aren't memory accounted, and they don't use predictable amount of memory
The backfill code has a few flaws.

1. It's not memory accounted when generating the SQL keys. Index keys can be large, and there can be many of them - both the `entries` list and the keys themselves in `backfill/backfill.go` should be accounted for.
2. The backfill code generates chunks of index entries by asking for 1000 rows at a time, and building entries for all 1000 rows. This is troublesome - what if the rows are really big? This will be mitigated in 20.1, since kv now takes a bytes limit. But, for inverted indexes, a single row can generate a ton of entries as well. The backfill code should act as an iterator - instead of generating all keys for all rows, it should generate keys for its batch of rows until the keys have hit a fixed-size buffer limit, write that out, then continue.

### PR DESCRIPTION
**Pull request by adityamaru**
This change introduces a new child "backfiller" monitor of the parent
bulk monitor.  This is used by both the index and column backfiller to
account for the memory usage of the row fetchers used in both.

Additionally, in the index backfiller, we account for the size of the
index keys, as well as the number of KVs, generated when building the
index entry for a chunk.

Closes: #46346

Release note: None

### CODE REVIEW

**Comment by adityamaru** [comment]
Follow up work on switching the index/column backfiller from a row count based `chunkSize` to a flat memory size based cluster setting is still to be done. I wanted to make a smaller, less intrusive change in case we decide to port this back to 20.2.

Also open to suggestions on how to test if there is any noticeable fallout with these new constraints.

**Comment by adityamaru** [comment]
The tests are failing on the first token allocation that the fetcher used by the index/column backfiller attempts, to scan rows. I haven't been able to pinpoint the issue yet. The "flow" memory monitor is used by all other processors and so I don't see why what we are plumbing through via the flowCtx is behaving any differently, still investigating.

File: pkg/sql/backfill/backfill.go
Comment (jordanlewis) [inline pkg/sql/backfill/backfill.go:617]: Hmm, this seems wrong to me - you've just made a large buffer, I'd expect you to be accounting for the memory of that buffer.

File: pkg/sql/backfill/backfill.go
Comment (jordanlewis) [inline pkg/sql/backfill/backfill.go:713]: It'd really be better to do this growing before you create the `buffer`. You in general want to grow the account and then allocate the memory if no errors.

**Review by jordanlewis — COMMENTED** [review]
Nice! The trickiest part of dealing with memory accounts is properly understanding their lifetimes, and making sure they're always cleared and closed. I'll leave that part of the review to David, who I expect would be more familiar with the lifetimes of these objects than I am.

**Comment by adityamaru** [comment]
TFTR! I just realized we have a bulk monitor and we probably want to create children of that considering the backfiller is a "bulk" operation. Refactoring and incorporating your feedback.

**Comment by adityamaru** [comment]
This is RFAL, added @pbardea too to get another set of eyes.

**Review by adityamaru — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/54156)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt, @jordanlewis, and @pbardea)

___
*[pkg/sql/backfill/backfill.go, line 617 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/54156#-MGp62mWByp66rTp3-_J-r1-617:-MHCPmqa8obwnzdS8sjb:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1de407078f023afe89bc0873990e01c07e4c42d3/pkg/sql/backfill/backfill.go#L617)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

Hmm, this seems wrong to me - you've just made a large buffer, I'd expect you to be accounting for the memory of that buffer.
</blockquote></details>

Done.
___
*[pkg/sql/backfill/backfill.go, line 713 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/54156#-MGp62mWByp66rTp3-_J-r1-713:-MHCPmTu5mbjTsakvp0c:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1de407078f023afe89bc0873990e01c07e4c42d3/pkg/sql/backfill/backfill.go#L713)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

It'd really be better to do this growing before you create the `buffer`. You in general want to grow the account and then allocate the memory if no errors.
</blockquote></details>

Done.
___
*[pkg/sql/backfill/backfill.go, line 35 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/54156#-MGp62mWByp66rTp3-_J-r2-35:-MHCPzOv6d7-fboCsFoE:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6af8513bd33ad19d43a67ff792a029df8951d224/pkg/sql/backfill/backfill.go#L35)):*
<details><summary><i>Previously, pbardea (Paul Bardea) wrote…</i></summary><blockquote>

Should this be `cockroachdb/errors`?
</blockquote></details>

Done.
___
*[pkg/sql/backfill/backfill.go, line 737 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/54156#-MGp62mWByp66rTp3-_J-r2-737:-MHCRkGJEF_0qPYCdsNj:b-vpmpqq) ([raw file](https://github.com/cockroachdb/cockroach/blob/6af8513bd33ad19d43a67ff792a029df8951d224/pkg/sql/backfill/backfill.go#L737)):*
<details><summary><i>Previously, pbardea (Paul Bardea) wrote…</i></summary><blockquote>

Do we want to attempt to grow the bound account even if `isBytesMonitored` is false? (Save above on 664 and 634.)
</blockquote></details>

cleaned it up!

File: pkg/sql/backfill/backfill.go
Comment (jordanlewis) [inline pkg/sql/backfill/backfill.go:631]: Would it be possible to always have a monitor here? I think all of these conditionals are a little bit unfortunate.

File: pkg/sql/backfill/backfill.go
Comment (jordanlewis) [inline pkg/sql/backfill/backfill.go:236]: I think that these calls will NPE if they haven't been initialized (like in the case you pass in a nil `mon`.  Should you check for `cb.mon == nil` before closing these things? What situations would you get a nil mon anyway, is it just tests?

**Review by adityamaru — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/54156)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt, @jordanlewis, and @pbardea)

___
*[pkg/sql/backfill/backfill.go, line 236 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/54156#-MGp62mWByp66rTp3-_J-r3-236:-MHGl3Fm34UbC43RgYub:bnm1aec) ([raw file](https://github.com/cockroachdb/cockroach/blob/0eb796382ecd3bb4b3e07b69f2dd230f73fd0115/pkg/sql/backfill/backfill.go#L236)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

I think that these calls will NPE if they haven't been initialized (like in the case you pass in a nil `mon`.  Should you check for `cb.mon == nil` before closing these things? What situations would you get a nil mon anyway, is it just tests?
</blockquote></details>

I don't think anyone should be passing a nil monitor. When used locally the planner's mon serves as the parent monitor, and when used via DistSQL processors the servers `BackfillerMon` is used as the parent monitor. This should be the case in tests as well.
___
*[pkg/sql/backfill/backfill.go, line 631 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/54156#-MGp62mWByp66rTp3-_J-r3-631:-MHGipaL6u7oPSKw1Q9J:bbb8ydo) ([raw file](https://github.com/cockroachdb/cockroach/blob/0eb796382ecd3bb4b3e07b69f2dd230f73fd0115/pkg/sql/backfill/backfill.go#L631)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

Would it be possible to always have a monitor here? I think all of these conditionals are a little bit unfortunate.
</blockquote></details>

The only case I was safeguarding against was if the planner used when running a backfill within a txn doesn't have a monitor associated with it. When the backfill is via DistSQL processors we always create a backfiller monitor and so this should always be true.
I believe the planner will always have a mon associated with it so I have removed the conditionals. I added some nil checks to the init method and a panic to the EncodeSecondaryIndexEntries method to enfroce that a mon is always passed.

**Review by jordanlewis — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/54156)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru, @dt, @jordanlewis, and @pbardea)

___
*[pkg/sql/backfill.go, line 1860 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/54156#-MHMrFPYEdlmJSX5BxH7:-MHMrFPYEdlmJSX5BxH8:b-ttf5gm) ([raw file](https://github.com/cockroachdb/cockroach/blob/402194016a727429196140e93f4e113595941b3f/pkg/sql/backfill.go#L1860)):*
> ```Go
> 	// This is the planner's memory monitor.
> 	if evalCtx.Mon != nil {
> 		indexBackfillerMon = execinfra.NewMonitor(ctx, evalCtx.Mon, "local-index-backfill-mon")
> ```

Isn't it going to be nil in this case?
___
*[pkg/sql/backfill/backfill.go, line 236 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/54156#-MGp62mWByp66rTp3-_J-r3-236:-MHMr9k01-sSRhx0G4yo:b-eq6xg6) ([raw file](https://github.com/cockroachdb/cockroach/blob/0eb796382ecd3bb4b3e07b69f2dd230f73fd0115/pkg/sql/backfill/backfill.go#L236)):*
<details><summary><i>Previously, adityamaru (Aditya Maru) wrote…</i></summary><blockquote>

I don't think anyone should be passing a nil monitor. When used locally the planner's mon serves as the parent monitor, and when used via DistSQL processors the servers `BackfillerMon` is used as the parent monitor. This should be the case in tests as well.
</blockquote></details>

That makes sense, but you have a case in which it stays `nil` if `evalctx.Mon` is nil, so it could happen, right?
___
*[pkg/sql/backfill/backfill.go, line 601 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/54156#-MHMqrYm3PnqrSUkSThS:-MHMqrYm3PnqrSUkSThT:b1u01qz) ([raw file](https://github.com/cockroachdb/cockroach/blob/402194016a727429196140e93f4e113595941b3f/pkg/sql/backfill/backfill.go#L601)):*
> ```Go
> 	// Create a bound account associated with the index backfiller monitor.
> 	if mon == nil {
> 		return errors.New("no memory monitor linked to IndexBacfiller during init")
> ```

Make this an `errors.NewAssertionError`

**Review by adityamaru — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/54156)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru, @dt, @jordanlewis, and @pbardea)

___
*[pkg/sql/backfill/backfill.go, line 236 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/54156#-MGp62mWByp66rTp3-_J-r3-236:-MHNqTU-8fuLlzsHWLkg:b-hq6p98) ([raw file](https://github.com/cockroachdb/cockroach/blob/0eb796382ecd3bb4b3e07b69f2dd230f73fd0115/pkg/sql/backfill/backfill.go#L236)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

That makes sense, but you have a case in which it stays `nil` if `evalctx.Mon` is nil, so it could happen, right?
</blockquote></details>

~~doh yes, you're right. Even if the backfiller init method errors out on seeing a nil memory monitor, the `defer ib/cb.Close()` in the `index/columnBackfillInTxn` will be called.~~ Added checks around closing the boundAccount and memMon, the fetcher is safe to close regardless as it has checks within its own Close method.

EDIT: sorry for the confusion, the defer call is after the InitForLocalUse. So the Close() methods should never see a nil monitor. I am happy to keep the nil check as a safeguard though.
___
*[pkg/sql/backfill/backfill.go, line 601 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/54156#-MHMqrYm3PnqrSUkSThS:-MHNqHNc-1s4gL-v6rgl:bdhy29g) ([raw file](https://github.com/cockroachdb/cockroach/blob/402194016a727429196140e93f4e113595941b3f/pkg/sql/backfill/backfill.go#L601)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

Make this an `errors.NewAssertionError`
</blockquote></details>

I could only find `errors.NewAssertionErrorWithWrappedf`. Since we don't have an err to wrap around I don't think we can use it?

**Comment by adityamaru** [comment]
sorry for the back and forth, but friendly ping 🙂

File: pkg/sql/backfill/backfill.go
Comment (pbardea) [inline pkg/sql/backfill/backfill.go:238]: If we're keeping this for a safeguard, should we also be checking that `boundAccount` is set?

**Review by pbardea — APPROVED** [review]
The lifecycle looks good to me. Would still be good to get an LGTM from Jordan too :)

**Review by jordanlewis — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/54156)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru, @dt, @jordanlewis, and @pbardea)

___
*[pkg/sql/backfill/backfill.go, line 601 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/54156#-MHMqrYm3PnqrSUkSThS:-MHlXalX1J_Tzu42rNRO:bfhkch7) ([raw file](https://github.com/cockroachdb/cockroach/blob/402194016a727429196140e93f4e113595941b3f/pkg/sql/backfill/backfill.go#L601)):*
<details><summary><i>Previously, adityamaru (Aditya Maru) wrote…</i></summary><blockquote>

I could only find `errors.NewAssertionErrorWithWrappedf`. Since we don't have an err to wrap around I don't think we can use it?
</blockquote></details>

It's called `errors.AssertionFailedf`, my bad. Use that instead of raw errors.New calls.
___
*[pkg/sql/backfill/backfill.go, line 73 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/54156#-MHlYE59-k-SSrASt-_i:-MHlYE59-k-SSrASt-_j:b1641yl) ([raw file](https://github.com/cockroachdb/cockroach/blob/1142439c7dd92120469cf3f4dd80a1968742403d/pkg/sql/backfill/backfill.go#L73)):*
> ```Go
> 
> 	mon          *mon.BytesMonitor
> 	boundAccount mon.BoundAccount
> ```

Add comments on these fields. The `mon` in particular - is this the column backfiller's mon, or a mon that's passed in from above? Important to make it clear to people what each field means in a struct when you can.

Also, I know the other fields in this struct are not commented. but that is no excuse not to do better going forward :grin: 
___
*[pkg/sql/backfill/backfill.go, line 509 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/54156#-MHlY9bI3FcsUw4uy-Zf:-MHlY9bI3FcsUw4uy-Zg:bbleh5h) ([raw file](https://github.com/cockroachdb/cockroach/blob/1142439c7dd92120469cf3f4dd80a1968742403d/pkg/sql/backfill/backfill.go#L509)):*
> ```Go
> func (ib *IndexBackfiller) Close(ctx context.Context) {
> 	ib.fetcher.Close(ctx)
> 	if ib.mon != nil {
> ```

I still don't totally understand why you need checks like this! But if you're sure they're necessary, fine by me.
___
*[pkg/sql/rowenc/index_encoding.go, line 1270 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/54156#-MHlY6Y3EuJATbZ2O4NN:-MHlY6Y3EuJATbZ2O4NO:b1elzej) ([raw file](https://github.com/cockroachdb/cockroach/blob/1142439c7dd92120469cf3f4dd80a1968742403d/pkg/sql/rowenc/index_encoding.go#L1270)):*
> ```Go
> 				return nil, errors.Wrap(err, "failed to allocate space for index keys")
> 			}
> 			fmt.Printf("key length %d\n", len(index.Key))
> ```

debug print

**Review by jordanlewis — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/54156)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru, @dt, @jordanlewis, and @pbardea)

___
*[pkg/sql/backfill/backfill.go, line 134 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/54156#-MHlYq-56hgn33yZXW46:-MHlYq-56hgn33yZXW47:bazzzb0) ([raw file](https://github.com/cockroachdb/cockroach/blob/1142439c7dd92120469cf3f4dd80a1968742403d/pkg/sql/backfill/backfill.go#L134)):*
> ```Go
> 	// Create a bound account associated with the column backfiller.
> 	if mon == nil {
> 		return errors.New("no memory monitor linked to ColumnBackfiller during init")
> ```

(this should be assertion error as well. there's a good reason for these by the way - it makes it so they turn into sentry reports when they bubble up! it's very much worth it to use these throughout the codebase.)

**Review by adityamaru — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/54156)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru, @dt, @jordanlewis, and @pbardea)

___
*[pkg/sql/backfill/backfill.go, line 601 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/54156#-MHMqrYm3PnqrSUkSThS:-MHlmF9j-NZCp8zDxHP4:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/402194016a727429196140e93f4e113595941b3f/pkg/sql/backfill/backfill.go#L601)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

It's called `errors.AssertionFailedf`, my bad. Use that instead of raw errors.New calls.
</blockquote></details>

Done.
___
*[pkg/sql/backfill/backfill.go, line 73 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/54156#-MHlYE59-k-SSrASt-_i:-MHln9UM3-IqECXTlMUz:b-7rl0jq) ([raw file](https://github.com/cockroachdb/cockroach/blob/1142439c7dd92120469cf3f4dd80a1968742403d/pkg/sql/backfill/backfill.go#L73)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

Add comments on these fields. The `mon` in particular - is this the column backfiller's mon, or a mon that's passed in from above? Important to make it clear to people what each field means in a struct when you can.

Also, I know the other fields in this struct are not commented. but that is no excuse not to do better going forward :grin: 
</blockquote></details>

done!
___
*[pkg/sql/backfill/backfill.go, line 134 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/54156#-MHlYq-56hgn33yZXW46:-MHlmHySCHDysoUKg_zN:b9widh0) ([raw file](https://github.com/cockroachdb/cockroach/blob/1142439c7dd92120469cf3f4dd80a1968742403d/pkg/sql/backfill/backfill.go#L134)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

(this should be assertion error as well. there's a good reason for these by the way - it makes it so they turn into sentry reports when they bubble up! it's very much worth it to use these throughout the codebase.)
</blockquote></details>

got it, thanks for the explanation
___
*[pkg/sql/backfill/backfill.go, line 238 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/54156#-MGp62mWByp66rTp3-_J-r5-238:-MHlqdpI47-QjFZ7YV_E:b-o4m1z) ([raw file](https://github.com/cockroachdb/cockroach/blob/1142439c7dd92120469cf3f4dd80a1968742403d/pkg/sql/backfill/backfill.go#L238)):*
<details><summary><i>Previously, pbardea (Paul Bardea) wrote…</i></summary><blockquote>

If we're keeping this for a safeguard, should we also be checking that `boundAccount` is set?
</blockquote></details>

boundAccounts Close() method has a check for a nil memory monitor associated with it. In that case it is just a no-op. boundAccount is not a pointer and so an NPE is not possible.
___
*[pkg/sql/rowenc/index_encoding.go, line 1270 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/54156#-MHlY6Y3EuJATbZ2O4NN:-MHllwsqAOtgG5wzvMWS:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1142439c7dd92120469cf3f4dd80a1968742403d/pkg/sql/rowenc/index_encoding.go#L1270)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

debug print
</blockquote></details>

Done.

**Review by jordanlewis — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/54156)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru, @dt, @jordanlewis, and @pbardea)

**Comment by adityamaru** [comment]
TFTR!

bors r=jordanlewis,pbardea

**Pull request merged**
